File: tests/c/epoll_pwait_zero_timeout_progress.c
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

/*
 * Regression guest for BOTH `epoll_pwait` scheduler defects in #1850. It takes
 * one optional argument selecting the scenario:
 *
 *   (none)     zero-timeout polling loop must not STARVE its producer.
 *   blocking   infinite-timeout wait must not DEADLOCK the scheduler.
 *
 * The two are different failure modes with different causes and they need
 * separate brackets. The zero-timeout case spins and is rescued by ordinary
 * timer preemption unless preemption is disabled, so its test must pass
 * `--max-timeslice=disabled` to detect anything. The blocking case is stuck
 * inside the kernel where no preemption can reach it, so it reproduces under
 * plain `--strict` and needs no special flag.
 *
 * The blocking scenario is the defect this PR is named for -- routing a
 * NULL-sigmask `epoll_pwait` away from inject-and-block. It was previously
 * unbracketed: reverting `handle_epoll_pwait` to its pre-#1850 body left the
 * entire suite green, including this file's zero-timeout tests, so nothing in
 * the repository detected the headline regression.
 *
 * A zero-timeout `epoll_pwait` cannot block, so it is tempting to inject it
 * straight into Linux with no scheduler interaction. Under sequentialized
 * threads that is a starvation bug: the polling thread never requests a
 * resource, so it is never descheduled between preemptions, and the worker
 * thread it is waiting on never gets the single logical CPU. `handle_poll`
 * has always taken a scheduler turn for zero-timeout calls for exactly this
 * reason; `handle_internal_epoll_pwait` must do the same.
 *
 * The main thread polls an eventfd with a zero timeout and spins until the
 * worker publishes. The worker does a little work, publishes a value, then
 * signals the eventfd. If the poller yields a turn, the worker runs and the
 * program prints its success line. If the poller monopolizes the CPU, this
 * hangs and the harness's `timeout` kills it (exit 124) -- which is the
 * regression signal, not a flaky failure.
 *
 * The syscall is issued directly rather than through glibc so the test pins
 * `epoll_pwait` with a NULL sigmask specifically: that is the exact routing
 * this regression concerns, and glibc is free to implement `epoll_wait` via
 * `epoll_pwait` or `epoll_pwait2` depending on version and kernel.
 *
 * Success is printing "epoll-pwait-zero-timeout-progress-ok <value>" and
 * exiting 0. Failure is a hang.
 */

#include <errno.h>
#include <pthread.h>
#include <stdatomic.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/epoll.h>
#include <sys/eventfd.h>
#include <sys/syscall.h>
#include <time.h>
#include <unistd.h>

#define PUBLISHED_VALUE 0x5eed1234ULL

static atomic_ullong g_value = 0;
static int g_event_fd = -1;

/* Raw epoll_pwait with an explicitly NULL sigmask. */
static int epoll_pwait_zero_timeout(int epfd, struct epoll_event* events, int maxevents) {
  return (int)syscall(SYS_epoll_pwait, epfd, events, maxevents, /*timeout=*/0, /*sigmask=*/NULL,
                      /*sigsetsize=*/(size_t)8);
}

static void* worker(void* arg) {
  (void)arg;
  /* A little work, so the poller genuinely has to yield rather than winning a
   * race by luck on the very first iteration. */
  unsigned long long acc = 0;
  for (unsigned long long i = 0; i < 100000ULL; i++) {
    acc += i;
  }
  atomic_store(&g_value, PUBLISHED_VALUE + (acc & 0ULL));

  uint64_t one = 1;
  if (write(g_event_fd, &one, sizeof(one)) != (ssize_t)sizeof(one)) {
    perror("write eventfd");
    _exit(1);
  }
  return NULL;
}

/* ---- Mode 2: the INFINITE-timeout deadlock, i.e. the defect #1850 is named
 * after. ----
 *
 * A blocking `epoll_pwait` that is injected raw and waited on holds the
 * scheduler turn inside the kernel, while the only thread that could satisfy
 * the wait sits in the run queue waiting for a turn that never comes. Nothing
 * breaks the cycle: unlike the zero-timeout case, timer preemption cannot
 * rescue it, because the waiter is blocked in the kernel rather than spinning.
 *
 * THE WORKER'S SLEEP IS LOAD-BEARING, not padding. Without it the worker wins
 * the race, writes before the main thread ever reaches `epoll_pwait`, and the
 * call returns immediately without blocking -- so the guest completes even with
 * the fix reverted and the test guards nothing. Measured while building this:
 * two earlier drafts without the sleep passed against the pre-#1850 handler.
 * Do not "simplify" it away.
 */
static void* sleeping_worker(void* arg) {
  (void)arg;
  struct timespec delay = {0, 50 * 1000 * 1000}; /* 50ms: the waiter must block first. */
  nanosleep(&delay, NULL);

  atomic_store(&g_value, PUBLISHED_VALUE);
  uint64_t one = 1;
  if (write(g_event_fd, &one, sizeof(one)) != (ssize_t)sizeof(one)) {
    perror("write eventfd");
    _exit(1);
  }
  return NULL;
}

static int run_blocking_mode(int epfd) {
  pthread_t tid;
  if (pthread_create(&tid, NULL, sleeping_worker, NULL) != 0) {
    perror("pthread_create");
    return 1;
  }

  /* Infinite timeout, NULL sigmask: the exact call `cmake` issues through
   * libuv, and the one that used to wedge the whole guest. */
  struct epoll_event got;
  for (;;) {
    int n = (int)syscall(SYS_epoll_pwait, epfd, &got, 1, /*timeout=*/-1, /*sigmask=*/NULL,
                         /*sigsetsize=*/(size_t)8);
    if (n < 0) {
      if (errno == EINTR) {
        continue;
      }
      perror("epoll_pwait(-1)");
      return 1;
    }
    if (n > 0) {
      break;
    }
  }

  uint64_t drained = 0;
  if (read(g_event_fd, &drained, sizeof(drained)) != (ssize_t)sizeof(drained)) {
    perror("read eventfd");
    return 1;
  }
  if (pthread_join(tid, NULL) != 0) {
    perror("pthread_join");
    return 1;
  }
  unsigned long long value = atomic_load(&g_value);
  if (value != PUBLISHED_VALUE) {
    fprintf(stderr, "unexpected published value: %llu\n", value);
    return 1;
  }
  printf("epoll-pwait-blocking-progress-ok %llu\n", value);
  return 0;
}

int main(int argc, char** argv) {
  const int blocking_mode = argc > 1 && strcmp(argv[1], "blocking") == 0;

  g_event_fd = eventfd(0, 0);
  if (g_event_fd < 0) {
    perror("eventfd");
    return 1;
  }

  int epfd = epoll_create1(0);
  if (epfd < 0) {
    perror("epoll_create1");
    return 1;
  }

  struct epoll_event ev;
  memset(&ev, 0, sizeof(ev));
  ev.events = EPOLLIN;
  ev.data.fd = g_event_fd;
  if (epoll_ctl(epfd, EPOLL_CTL_ADD, g_event_fd, &ev) != 0) {
    perror("epoll_ctl");
    return 1;
  }

  if (blocking_mode) {
    int rc = run_blocking_mode(epfd);
    close(epfd);
    close(g_event_fd);
    return rc;
  }

  pthread_t tid;
  if (pthread_create(&tid, NULL, worker, NULL) != 0) {
    perror("pthread_create");
    return 1;
  }

  /* Poll with a zero timeout until the worker publishes. No sleep, no yield:
   * the scheduler turn taken inside the zero-timeout epoll_pwait path is the
   * only thing that lets the worker run. */
  struct epoll_event got;
  for (;;) {
    int n = epoll_pwait_zero_timeout(epfd, &got, 1);
    if (n < 0) {
      if (errno == EINTR) {
        continue;
      }
      perror("epoll_pwait");
      return 1;
    }
    if (n > 0) {
      break;
    }
  }

  uint64_t drained = 0;
  if (read(g_event_fd, &drained, sizeof(drained)) != (ssize_t)sizeof(drained)) {
    perror("read eventfd");
    return 1;
  }

  if (pthread_join(tid, NULL) != 0) {
    perror("pthread_join");
    return 1;
  }

  unsigned long long value = atomic_load(&g_value);
  if (value != PUBLISHED_VALUE) {
    fprintf(stderr, "unexpected published value: %llu\n", value);
    return 1;
  }

  printf("epoll-pwait-zero-timeout-progress-ok %llu\n", value);
  close(epfd);
  close(g_event_fd);
  return 0;
}

File: hermit-cli/tests/epoll_pwait_zero_timeout_progress.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! Regression test: a zero-timeout `epoll_pwait` polling loop must not starve
//! its producer under ordinary sequentialized strict execution.
//!
//! A zero-timeout `epoll_pwait` cannot block, so injecting it straight into
//! Linux looks free. Under sequentialized threads it is not: a poller that
//! never requests a resource is never descheduled between preemptions, and the
//! worker it is polling for never gets the single logical CPU. `handle_poll`
//! has always taken a scheduler turn for zero-timeout calls, and the
//! record/replay arm of `handle_epoll_pwait` does too; the plain-strict path
//! must match. Before the fix this guest hangs.
//!
//! This asserts PROGRESS, not a return value: a syscall-return assertion
//! passes happily while the producer starves. The failure signal is `timeout`
//! killing the run (exit 124).

use std::fs;
use std::path::Path;
use std::path::PathBuf;
use std::process::Command;
use std::sync::OnceLock;

/// A healthy run finishes in well under a second. A starved run spins forever,
/// so this bound only has to be generous enough to never fire on a loaded box.
const TIMEOUT_SECONDS: u64 = 60;

const SUCCESS_MARKER: &str = "epoll-pwait-zero-timeout-progress-ok";

static GUEST: OnceLock<PathBuf> = OnceLock::new();

fn guest() -> &'static Path {
    GUEST
        .get_or_init(|| {
            let repository = Path::new(env!("CARGO_MANIFEST_DIR"))
                .parent()
                .expect("hermit-cli should be inside the repository");
            let build_root =
                Path::new(env!("CARGO_TARGET_TMPDIR")).join("epoll-pwait-zero-timeout-progress");
            fs::create_dir_all(&build_root).expect("failed to create build directory");
            let output = build_root.join("epoll_pwait_zero_timeout_progress");
            let mut command = Command::new("cc");
            command
                .args([
                    "-std=c11",
                    "-O2",
                    "-g",
                    "-pthread",
                    "-D_GNU_SOURCE",
                    "-Wall",
                    "-Wextra",
                    "-Werror",
                ])
                .arg(repository.join("tests/c/epoll_pwait_zero_timeout_progress.c"))
                .arg("-o")
                .arg(&output);
            let status = command
                .status()
                .expect("failed to run cc to build epoll_pwait_zero_timeout_progress guest");
            assert!(status.success(), "guest compilation failed: {command:?}");
            output
        })
        .as_path()
}

/// Run the guest under ordinary sequentialized strict mode -- the exact
/// configuration that routes a NULL-sigmask `epoll_pwait` into
/// `handle_internal_epoll_pwait`. Deliberately no `--chaos` and no
/// record/replay: those arms already took a scheduler turn before the fix, so
/// exercising them would not detect the regression.
fn run_plain_strict(extra: &[&str]) {
    run_guest(extra, &[], SUCCESS_MARKER);
}

/// Shared driver. `extra` are hermit flags; `guest_args` select the scenario
/// inside the guest; `marker` is the success line that scenario prints.
fn run_guest(extra: &[&str], guest_args: &[&str], marker: &str) {
    let mut command = Command::new("timeout");
    command
        .arg("--kill-after=2s")
        .arg(format!("{TIMEOUT_SECONDS}s"))
        .arg(env!("CARGO_BIN_EXE_hermit"))
        .args([
            "run",
            "--base-env=minimal",
            "--no-virtualize-cpuid",
            "--strict",
        ])
        .args(extra)
        .arg("--")
        .arg(guest())
        .args(guest_args);

    let rendered = format!("{command:?}");
    let output = command
        .output()
        .unwrap_or_else(|error| panic!("failed to start guest: {rendered}: {error}"));

    // `timeout` exits 124 when it has to kill the child. That is precisely the
    // starvation this test guards: the poller monopolized the CPU and the
    // worker never ran.
    assert_ne!(
        output.status.code(),
        Some(124),
        "zero-timeout epoll_pwait starved its producer (timed out): {rendered}\nstdout:\n{}\nstderr:\n{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr),
    );
    assert!(
        output.status.success(),
        "guest failed: {rendered}\nstatus: {}\nstdout:\n{}\nstderr:\n{}",
        output.status,
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr),
    );

    let stdout = String::from_utf8(output.stdout).expect("guest stdout should be UTF-8");
    assert!(
        stdout.contains(marker),
        "guest did not report progress: {rendered}\nstdout:\n{stdout}"
    );
}

/// THE REGRESSION DETECTOR. Timer preemption is disabled, which removes the
/// only other mechanism that can rescue a non-yielding poller, so the
/// scheduler turn taken inside the zero-timeout `epoll_pwait` path is the sole
/// reason this guest can finish.
///
/// Measured both ways against this exact guest: with the scheduler request
/// removed from `handle_internal_epoll_pwait`, this test FAILS with the guest
/// killed at 60s (`timeout` exit 124); with the request restored it passes in
/// well under a second. That bracket is what makes it a guard rather than a
/// hopeful assertion.
#[test]
fn zero_timeout_epoll_pwait_yields_without_timer_preemption() {
    run_plain_strict(&["--max-timeslice=disabled"]);
}

/// Smoke coverage of the ordinary configuration, with timer preemption left
/// on. Kept because it exercises the common path end to end.
///
/// DO NOT MISTAKE THIS FOR A REGRESSION GUARD. Measured: with the scheduler
/// request removed it still PASSES, because ordinary timer preemption
/// deschedules the poller often enough for the worker to run. It cannot detect
/// the bug it appears to be about. The detector is
/// `zero_timeout_epoll_pwait_yields_without_timer_preemption` above; if that
/// one is ever weakened or deleted, this test does not cover for it.
#[test]
fn zero_timeout_epoll_pwait_ordinary_strict_smoke() {
    run_plain_strict(&[]);
}

/// THE BRACKET FOR THE HEADLINE FIX, and the reason this file exists in its
/// current shape.
///
/// An infinite-timeout `epoll_pwait` that is injected raw and waited on holds
/// the scheduler turn inside the kernel while the only thread that could
/// satisfy it waits for a turn that never comes. That deadlock is the defect
/// #1850 is named after, and until this test it was the ONE change in the PR
/// with no guard: reverting `handle_epoll_pwait` to its exact pre-#1850 body
/// left everything green -- both zero-timeout tests in this file 2/2,
/// `epoll_determinism` 7/7, and `hermit-detcore --lib` 388/388.
///
/// Measured both ways against this exact guest, plain `--strict`, ptrace:
/// with the pre-#1850 handler the guest is killed at 60s (`timeout` exit 124);
/// at this head it exits 0. No flag is needed to expose it, unlike the
/// zero-timeout case -- a thread blocked in the kernel is beyond the reach of
/// timer preemption, which is exactly why this failure is a hard hang rather
/// than starvation.
#[test]
fn blocking_epoll_pwait_does_not_deadlock_the_scheduler() {
    run_guest(&[], &["blocking"], "epoll-pwait-blocking-progress-ok");
}
